### PR TITLE
Baseline classifier from embeddings dataset

### LINKED ISSUE #84: Build a baseline classifier using mixit/birdnet embeddings.
Copy pasting from slack:

You can sync gs://birdclef-2023/data/processed/birdclef-2023/train_embeddings_consolidated_v1/ to your local directory and use spark to read process it.

```                                                                                
root
 |-- embedding: array (nullable = true)
 |    |-- element: double (containsNull = true)
 |-- predictions: array (nullable = true)
 |    |-- element: map (containsNull = true)
 |    |    |-- key: string
 |    |    |-- value: long (valueContainsNull = true)
 |-- track_name: string (nullable = true)
```

It has 6.6k distinct tracks, and it also has the tracks with mixit. Turns out i also duplicated the work of running birdnet stuff on the mp3 and the wav files (meh). This is missing about 10k of tracks, so I have to figure out what went wrong and also debug a few more things. This should be enough build a classifier on. It's a 10gb dataset, so it might be ~2gb in memory if you just keep a single channel. You might consider using spark for classification (https://spark.apache.org/docs/latest/ml-classification-regression.html#logistic-regression), or dumping out a derived dataset thats flatter so you can use it with another tool of your choice.
This also includes a bunch of non-birdcalls too, so you'll have to filter it so it only contains confident results. 

The pipeline should probably do something like the following in spark:

- filter down to the source track, or to the first channel of the mixit model
- explode predictions array
- apply sigmoid function to predictions, and keep the tracks that are birdcalls (50%?)
- parse the species from the track name
- save the dataframe with the embedding, the species, and the confidence

Then with this dataset, you should be able to feed this into sklearn, pyspark, pytorch, etc. Treat this as a multi-class prediction. Let me know if you have trouble starting on this, but this will end up being one of our baselines models probably.

### PR DESCRIPTION
**Pull request by nzhongathan**

### CODE REVIEW

File: notebooks/nzh-20230417-00-embeddings-classifier.ipynb
Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    import pyspark.sql.functions as f
</pre><p>FYI, I usually do single letter module names with a capital. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1178446740'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.                                         explode_preds.col["label"].alias("label"),
</pre><p>The shorthand to get the individual columns is to alias the previous exploded field (<code>F.explode(...).alias("col")</code>, and to do <code>"col.*"</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1178446745'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.                         (1 / (1 + f.exp(-1 * explode_preds.probability))).alias("probability"))
</pre><p>You should be able to use the pre-computed confidence field, which should be the sigmoid of this probability field.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1178446746'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='9'/>

Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    print(len(species))
</pre><p>I'd also recommend plotting a bar chart of the count per species. There might be skew between the different classes.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1178446748'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    print(treeClassifier.score(test_x, test_y))
</pre><p>This looks good, you might consider printing a few other metrics, like precision/recall/f1-score etc. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1178446749'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='18'/>

Comment (nzhongathan) [reply notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: I don't think there is a confidence field in the predictions map. There is only rank, index, label, mapped_label, probability

Comment (nzhongathan) [reply notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: And if I take the sigmoid of the sigmoid of the probabilities, all my values are between 0.5 and 0.8 (there is bar graph in the notebook). Not sure if this was correct since that would mean all tracks contain bird calls?

Comment (nzhongathan) [reply notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: Should I test with other, more complex classifiers? metrics seemed pretty solid so I didn't know if we should just run with this or try other things.

Comment (acmiyaguchi) [reply notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: Simple is good, but we can certainly have any number of simple models if we want to try them. Gradient boosted decision trees are about as complex as you want to go, I'd say logistic regression is something we'd also want to include for sure. If you do use a more complex model, try out a simple grid search over parameters.

Comment (acmiyaguchi) [reply notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: I must have mixed this up with the v1 dataset, this is fine then.

Comment (nzhongathan) [reply notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: added logistic regression and a grid search on xgboostclassifier... the metrics on one of the species looks really poor probably due to a lack of training data and the logistic regression seems to perform the best so far

File: notebooks/nzh-20230417-00-embeddings-classifier.ipynb
Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.            for i in range(3337)
</pre><p>I just realized that there's a bug in my code, the embedding size should ~300 and not ~3000. Thanks for helping me realize this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1181293897'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='15'/>

Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #23.            precision_score(truth, preds, average="macro"),
</pre><p>I think you made some copy paste errors here, the recall and f1_score lines are averaging over the precision.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1181293898'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='17'/>

**Comment by acmiyaguchi** [comment]
The notebook overall seems reasonable to me. I'd also play around with serializing the model to disk (via pickle or joblib), and maybe think about how you would write inference code. This is also a good time to convert this into a luigi script now that you have the basic workflow down.

File: models/logistic_reg_3.pkl
Comment (acmiyaguchi) [inline models/logistic_reg_3.pkl:1]: I'd actually prefer not to check this in here, but to put them in gcs under `data/models/baseline` or the like.

File: scripts/embedding_classifier.py
Comment (acmiyaguchi) [inline scripts/embedding_classifier.py:1]: This approach won't work super well for the competition, mostly because this assumes that the input data is in the parquet format (which we could try out if we modified the preprocessing scripts I suppose). The input for this kind of thing would actually be a directory to all of the ogg files of 10m lengths, so the inference script would need to run birdnet, then run the baseline classifier, and then do some sort of averaging across overlapping samples so that we have a classification per 5-second branch.

File: scripts/embedding_classifier.py
Comment (acmiyaguchi) [inline scripts/embedding_classifier.py:64]: This is the schema of the submission csv, fwiw:

https://www.kaggle.com/competitions/birdclef-2023/data?select=sample_submission.csv

File: scripts/embedding_classifier.py
Comment (nzhongathan) [reply scripts/embedding_classifier.py:1]: do you know how i can run birdnet through a command? or will kaggle have access to luigi so i could just run the luigi script

Comment (acmiyaguchi) [reply scripts/embedding_classifier.py:1]: You'd probably use it in a script something like this: https://github.com/dsgt-birdclef/birdclef-2023/blob/77c96d306b6155c5698b462949bd089764d19600/workflows/build_train_dataset.py#L94-L132

I'll probably shift my focus in the next week for getting something running on Kaggle, but I have a pretty good idea of how to do this. We have access to the shell and general python stuff in the notebooks there, since we just get dropped into a notebook running on docker.

File: notebooks/nzh-20230417-00-embeddings-classifier.ipynb
Comment (acmiyaguchi) [inline notebooks/nzh-20230417-00-embeddings-classifier.ipynb:1]: It's usually easiest to do this kind of thing by creating a dataframe, and then by using <code>pd.write_csv.</code> I would do something where I zip the probabilities and the classes together, and then create a dictionary from that zipped list. 

Something along these lines

```python
res = []
for row in rows:
    assert len(species) == len(predictions), "this only works when the two lists are the same dim"
    d = dict(*zip(species, predictions))
    d["row_id"] = row
    res.append(d)
res_df = pd.DataFrame(res)
```


 _Reply via <a href='https://app.reviewnb.com/dsgt-birdclef/birdclef-2023/pull/86/discussion/#comment-1188180521'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='27'/>

**Comment by acmiyaguchi** [comment]
I'd go ahead and merge this PR when you feel like you're ready to move on to either building out a luigi script or using the larger dataset. The initial classification results are not too bad actually.

This notebook might be useful for getting data in the right shape, or for doing some other manipulation that might be a bit awkward. https://github.com/dsgt-birdclef/birdclef-2023/blob/main/notebooks/acm-20230508-00-embedding-ideas.ipynb

**Comment by nzhongathan** [comment]
I can merge this in soon but I also have the script file under this PR as well so should I just wait until after the script to merge the PR in or make a new PR with just the notebook?

**Comment by acmiyaguchi** [comment]
Either way is fine; I just wouldn't increase scope too much on this PR. I'm assuming that testing against the larger embedding dataset will be done in another notebook, so that should come in a follow up.

**Review by acmiyaguchi — APPROVED** [review]

**Pull request merged**
